### scRNAbase: Pre-Processing Toolkit for scRNA-Seq

<br />



<p align="right">
<img  src="https://github.com/jkubis96/Logos/blob/main/logos/jbs_current.png?raw=true" alt="drawing" width="180" />
</p>


### Author: Jakub Kubiś

<div align="left">
 Institute of Bioorganic Chemistry<br />
 Polish Academy of Sciences<br />
 Laboratory of Single Cell Analyses<br />

<br />

<p align="left">
<img  src="fig/lsca.png" alt="drawing" width="250" />
</p>
</div>


<br />


## Description

A comprehensive toolkit for the basic analysis of single-cell RNA sequencing (scRNA-seq) data without cell barcodes, relying exclusively on unique molecular identifiers (UMIs). Developed at the Laboratory of Single-Cell Analyses, Institute of Bioorganic Chemistry, PAS, in Poznań, Poland.


Core functionalities include:
- Genome Downloading
- Reads Quality Control - [fastp](https://github.com/OpenGene/fastp)
- UMI Selection - [UMI-tools](https://github.com/CGATOxford/UMI-tools)
- Genome Annotation - [STAR](https://github.com/alexdobin/STAR)
- Annotation Adjustment (names repairing, UTR extending) - [GTF-tool](https://github.com/jkubis96/GTF-tool)
- Reads Mapping - [STAR](https://github.com/alexdobin/STAR)
- BAM Indexing - [samtools](https://github.com/samtools/samtools)
- Features selection - [featureCounts](https://github.com/ShiLab-Bioinformatics/subread)

scRNAbase enables researchers to perform essential pre-processing analyses on scRNAseq datasets in a simple and efficient way.



#### Installation - docker pull

```
docker pull jkubis96/scrna_base:1.0.5
```

<br />



#### Genome directories hierarchy

genome/ ................................................... Main directory
├── Homo_sapiens/ ............................... Subdirectory for Homo sapiens genome
│   └── index/ ............................................Subfolder for indexes and annotations
│       ├── 100/ ..............................................Index for read length 100
│       └── 150/ ..............................................Index for read length 150
├── Mus_musculus/ ...............................Subdirectory for Mus musculus genome
│   └── index/ ...........................................Subfolder for indexes and annotations
│       ├── 100/ ..............................................Index for read length 100
│       └── 150/ ..............................................Index for read length 150

<br />


#### Genome Downloading

genome_downloading - download genome and annotation files

Options for genome_downloading:
  --genome_link (URL) - URL to download the genome file (required)
  --annotation_link (URL) - URL to download the annotation file (required)
  --species (string) - name of the species (e.g., Homo_sapiens) (required)

<br />


```
cd genome
docker run -it -v $(pwd):/app/genome jkubis96/scrna_base:1.0.5 bash -c "cd /app/genome && bash"

Example:

IndexPip genome_downloading --genome_link URL --annotation_link URL --species Homo_sapiens
```

<br />


#### Genome Annotation

 genome_indexing - create reference files and index the genome


Options for genome_indexing:
  --species (string) - name of the species (e.g., Homo_sapiens) (required)
  --reads_length (int) - length of reads for STAR index (required)
  --CPU (int) - number of threads to use (optional, default: number of CPU cores - 2)
  --memory (int) - amount of memory for STAR (optional, default: all available RAM)

Additional parameters:
  --optimize <bool>             Run GTF/GFF3 file adjustment (optional, default: TRUE)
  --extend <bool>               Extend parameter for genome prep (optional, default: FALSE)
  --five_prime_utr <int>        Length of 5' UTR (optional, default: 400)
  --three_prime_utr <int>       Length of 3' UTR (optional, default: 1000)
  --coding_elements <list>      List of coding annotation elements (optional, default: EXON,CDS,TRANSCRIPT,MRNA)
  --space <int>                 Minimal differential factor for separating features [genes] (optional, default: 100000)

<br />


```
cd genome
docker run -it -v $(pwd):/app/genome jkubis96/scrna_base:1.0.5 bash -c "cd /app/genome && bash"

Example:

IndexPip genome_indexing --species Homo_sapiens --reads_length 100 --optimize TRUE --extend TRUE
```



<br />


#### Fastq - experimental data directories

data/ ......................................................Main directory for experimental FASTQ data
├── name.1_R1.fastq.gz .....................Read 1 of sample 1 in FASTQ format (compressed)
├── name.1_R2.fastq.gz .....................Read 2 of sample 1 in FASTQ format (compressed)
├── name.2_R1.fastq.gz .....................Read 1 of sample 2 in FASTQ format (compressed)
├── name.2_R2.fastq.gz .....................Read 2 of sample 2 in FASTQ format (compressed)
├── results/ ..........................................Directory for analysis results
│   └── name.1.fastp_report.html ..........Quality control report for sample 1 (generated by fastp)
│   └── name.2.fastp_report.html ..........Quality control report for sample 2 (generated by fastp)
│   └── matrices/ ....................................Subdirectory for count matrices
│       ├── gene_id_exon_name.1_genes_count_matrix ....Count matrix  exon sample 1
│       ├── gene_id_UTR_name.1_genes_count_matrix .....Count matrix  UTR sample 1
│       ├── gene_id_exon_name.2_genes_count_matrix ....Count matrix  exon sample 2
│       ├── gene_id_UTR_name.2_genes_count_matrix .....Count matrix  UTR sample 2


<br />

#### Reads Mapping

matrices_creating - function processes FASTQ files to create gene and transcript count matrices. It includes steps such as read trimming, UMI extraction, mapping, deduplication, and matrix creation.

Options:
  --umi_length <int>              Length of the Unique Molecular Identifier (UMI). Required.
  --reads_length <int>            Read length for the input sequencing reads. Required.
  --species <string>              Species name (e.g., Homo_sapiens, Mus_musculus). Required.
  --qc_reads <TRUE|FALSE>         Whether to perform quality control of reads (TRUE by default).
  --multi <TRUE|FALSE>            Count reads mapped to multiple locations (FALSE by default).
  --annotation_names <string>     Names of annotations to include, provided as a list. Defaults to 'gene_id,gene_name'.
  --annotation_side <string>      Specifies the side of the annotation, e.g., 'exon,intron,UTR,five_prime_UTR,three_prime_UTR,transcript'. 
                                   Defaults to 'exon,transcript,three_prime_UTR,five_prime_UTR,CDS'.
  --CPU <int>                     Number of CPU threads to use. Defaults to (total CPU cores - 2).



Note:
  - All required arguments (--umi_length, --reads_length, --species) must be provided.
  - Ensure that the genome directory and annotation files are correctly set up using the IndexingPip function.

<br />

```
cd data
docker run -it -v $(pwd):/data -v genome:/app/genome jkubis96/scrna_base:1.0.5

Example:

CountFeatures matrices_creating --umi_length 12 --reads_length 100 --species Homo_sapiens --qc_reads TRUE --annotation_names gene_id,gene_name --annotation_side exon,intron,UTR,five_prime_UTR,three_prime_UTR,transcript
```
<br />

<br />


#### Have fun JBS©